File: backend/app/schemas/terms.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.core.currencies import CURRENCIES
from app.schemas.cards import HandoverMethod, PaymentMethod

#: T3.12.04 — what used to be the «Груз» section and belongs to the cargo now.
#: `D-CARGO-MODEL`: the cargo is written once at the response and does not
#: change, so terms that carried their own weight or their own «хрупкое» would be
#: a second copy of the parcel that two people could edit. Named here so that a
#: client still sending them is refused by name, not silently ignored.
CARGO_FIELDS: frozenset[str] = frozenset(
    {
        "weight_kg",
        "dimensions_cm",
        "declared_value",
        "cargo_what",
        "cargo_packaging",
        "cargo_fragile",
        "cargo_open_on_handover",
        "cargo_url",
    }
)


class TermsIn(BaseModel):
    """A proposal from either side.

    `description` is the only free-text field and it goes to the encrypted
    column, not into the payload — the payload holds what the server has to be
    able to read (§6.9.3).

    T3.12.04 — no cargo. The weight and dimensions the price is normalised by
    are read from the deal's cargo (`api.terms._build_payload`).
    """

    price_total: float = Field(gt=0)
    # T3.11.07 — four characters, because `USDT` and `USDC` are four. A deal is
    # negotiated in the currency the trip was published in, so a limit narrower
    # than the trip's would refuse the ordinary continuation of that trip.
    currency: str = Field(default="USD", min_length=3, max_length=4)
    deadline: datetime | None = None
    payment_method: PaymentMethod = "cash_on_delivery"
    description: str | None = None
    # Set when countering: the proposal this one replaces.
    supersedes_id: uuid.UUID | None = None

    # ── T3.11.27 — the whole agreement, in sections ───────────────────────
    #
    # Owner's decision 2026-09-07: one card instead of four. The same facts used
    # to be spread across `terms.*`, `handover.conditions`, `pickup.proposed`
    # and `dropoff.proposed`, and «договориться о передаче» stood as a separate
    # step *after* both sides had agreed the deal — describing an agreement they
    # had already reached.
    #
    # The fields are flat with section prefixes rather than nested: the sections
    # exist to name what changed («изменены условия: Передача, Оплата»).
    # `SECTION_OF` in `core.cards` is what maps one to the other.
    #
    # All of them optional: a half-filled agreement is the normal state of the
    # first stage rather than an error.
    handover_method: HandoverMethod | None = None
    handover_place: str | None = Field(default=None, max_length=200)
    handover_at: datetime | None = None

    delivery_method: HandoverMethod | None = None
    delivery_place: str | None = Field(default=None, max_length=200)
    delivery_at: datetime | None = None

    #: T3.11.27 — who pays the carrier. It decides whose button closes the deal:
    #: «получил» and «рассчитался» are one press for the person who does both,
    #: and two presses when the recipient is somebody who owes nothing.
    payer: Literal["sender", "recipient"] = "sender"

    #: Sections the **carrier** has locked in this deal. A locked section is
    #: theirs to edit; everything else is open to both. Per deal rather than per
    #: trip (owner's decision 2026-09-07): a carrier is stricter with one sender
    #: and softer with another.
    locked: list[str] = Field(default_factory=list)

    @model_validator(mode="before")
    @classmethod
    def _no_cargo(cls, data: Any) -> Any:
        if isinstance(data, dict):
            sent = sorted(CARGO_FIELDS & set(data))
            if sent:
                raise ValueError(
                    f"The cargo is not part of the terms: {', '.join(sent)}"
                )
        return data

    @field_validator("locked")
    @classmethod
    def _known_sections(cls, v: list[str]) -> list[str]:
        from app.core.cards import DEAL_SECTIONS

        unknown = set(v) - set(DEAL_SECTIONS)
        if unknown:
            raise ValueError(f"unknown sections: {sorted(unknown)}")
        return list(dict.fromkeys(v))

    @field_validator("currency")
    @classmethod
    def _known_currency(cls, v: str) -> str:
        code = v.strip().upper()
        if code not in CURRENCIES:
            raise ValueError(f"unknown currency: {code}")
        return code


class TermsOut(BaseModel):
    id: uuid.UUID
    deal_id: uuid.UUID
    card_kind: str
    card_state: str
    requires_ack_by: str | None
    supersedes_id: uuid.UUID | None
    payload: dict[str, Any]
    description: str | None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_message(cls, msg) -> "TermsOut":
        return cls(
            id=msg.id,
            deal_id=msg.deal_id,
            card_kind=msg.card_kind,
            card_state=msg.card_state.value if msg.card_state else "pending",
            requires_ack_by=(
                msg.requires_ack_by.value if msg.requires_ack_by else None
            ),
            supersedes_id=msg.supersedes_id,
            payload=msg.card_payload or {},
            description=msg.text,
            created_at=msg.created_at,
        )

File: frontend/src/api/terms.ts
import api from './client'
import type { VaultMessage } from './dealvault'

export interface NormalizedTerms {
  direction: string | null
  route: string
  distance_km: number | null
  weight_kg: number
  chargeable_weight_kg: number
  price_total: number
  currency: string
  price_per_kg: number | null
  price_per_km: number | null
}

export interface TermsPayload {
  weight_kg?: number
  dimensions_cm?: number[] | null
  declared_value?: number
  price_total?: number
  currency?: string
  deadline?: string | null
  payment_method?: string
  normalized?: NormalizedTerms
  below_carrier_minimum?: boolean
  agreed_at?: string
  proposal_id?: string
  platform_params?: Record<string, string>
  // T3.11.27 — the rest of the agreement: one card holds what four used to.
  cargo_what?: string | null
  cargo_packaging?: string | null
  cargo_fragile?: boolean
  cargo_open_on_handover?: boolean
  cargo_url?: string | null
  handover_method?: string | null
  handover_place?: string | null
  handover_at?: string | null
  delivery_method?: string | null
  delivery_place?: string | null
  delivery_at?: string | null
  /** Who pays the carrier — it decides whose button closes the deal. */
  payer?: 'sender' | 'recipient'
  /** Sections the carrier has locked in this deal. */
  locked?: string[]
}

export interface Terms {
  id: string
  deal_id: string
  card_kind: string
  card_state: string
  requires_ack_by: string | null
  supersedes_id: string | null
  payload: TermsPayload
  description: string | null
  created_at: string
}

export interface TermsInput {
  weight_kg: number
  price_total: number
  declared_value: number
  currency?: string
  dimensions_cm?: number[] | null
  deadline?: string | null
  /** T3.11.27 — one of `PAYMENT_METHODS`: the same three words the trip is
   *  published with. Required in practice — the owner's rule 2026-09-08 is
   *  «модель расчёта обязательна» — but optional on the wire so an older client
   *  and the server's own default stay in step. */
  payment_method?: string
  description?: string | null
  supersedes_id?: string | null
  // T3.11.27 — the four sections. Every one optional: a deal born from the
  // board form has a price and a weight and nothing else, and a half-filled
  // agreement is the normal state of the first stage.
  cargo_what?: string | null
  cargo_packaging?: string | null
  cargo_fragile?: boolean
  cargo_open_on_handover?: boolean
  cargo_url?: string | null
  handover_method?: string | null
  handover_place?: string | null
  handover_at?: string | null
  delivery_method?: string | null
  delivery_place?: string | null
  delivery_at?: string | null
  payer?: 'sender' | 'recipient'
  locked?: string[]
}

/** Current contract, or the proposal still awaiting an answer. `null` when
 *  neither exists yet. */
export async function getTerms(dealId: string): Promise<Terms | null> {
  const { data } = await api.get<Terms | null>(`/api/deals/${dealId}/terms`)
  return data
}

export async function proposeTerms(dealId: string, input: TermsInput): Promise<Terms> {
  const { data } = await api.post<Terms>(`/api/deals/${dealId}/terms`, input)
  return data
}

/** T3.36–T3.39 — every other card goes through one endpoint. */
export async function raiseCard(
  dealId: string,
  kind: string,
  payload: Record<string, unknown> = {},
  text?: string,
): Promise<VaultMessage> {
  const { data } = await api.post<VaultMessage>(`/api/deals/${dealId}/cards`, {
    kind,
    payload,
    text,
  })
  return data
}

/** T3.11.27 — a declaration and the photographs that prove it, in one request.
 *
 *  «Без фото карточка в чат добавляться не должна» (owner, 2026-09-12). Two
 *  requests could not promise that: the card was committed first, and a refused
 *  upload then left a declaration in an append-only chain that nobody could
 *  confirm and nobody could take back. The server validates every file before
 *  it writes anything, and one transaction covers both.
 *
 *  The kind of attachment is not passed: the card's own spec declares what
 *  evidence it requires, and a client naming a different one could only be
 *  wrong.
 */
export async function raiseCardWithFiles(
  dealId: string,
  kind: string,
  files: File[],
  payload: Record<string, unknown> = {},
  text?: string,
): Promise<VaultMessage> {
  const form = new FormData()
  form.append('kind', kind)
  form.append('payload', JSON.stringify(payload))
  if (text) form.append('text', text)
  for (const file of files) form.append('files', file)
  const { data } = await api.post<VaultMessage>(
    `/api/deals/${dealId}/cards/with-files`,
    form,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  )
  return data
}

/** T3.11.27 — «я сейчас правлю», for the next two minutes.
 *
 *  Taken before the change and released by it. While it is open the other side
 *  neither edits nor confirms the agreement, so what they read next is a
 *  finished version rather than somebody's second thoughts. `409` means the
 *  other side got there first — «запускай ещё раз».
 */
export const takeEditHold = (dealId: string) =>
  api.post<{ until: string }>(`/api/deals/${dealId}/terms/hold`)
